javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -22,24 +22,23 @@ class ArrayDeque[E] private (private val inner: js.Array[E])
   self =>
 
   private var status = 0
+  private var startIndex = 0 // inclusive, 0 <= startIndex < inner.length
+  private var endIndex = inner.length // exclusive, 0 < endIndex <= inner.length
+  private var empty = true
 
-  def this(initialCapacity: Int) = {
-    this(new js.Array[E])
-
-    if (initialCapacity < 0)
-      throw new IllegalArgumentException
-  }
+  def this(initialCapacity: Int) =
+    this(new js.Array[E](Math.max(initialCapacity, 16)))
```

Comment:
Question: this (admittedly old) blog post seems to suggest that even if a newly created array is immediately filled with elements, it will still be treated as sparse by the runtime (due to its sparse construction). Is that an issue here?

```js
const array = new Array(3);
// The array is sparse at this point, so it gets marked as
// `HOLEY_SMI_ELEMENTS`, i.e. the most specific possibility given
// the current information.
array[0] = 'a';
// Hold up, that’s a string instead of a small integer… So the kind
// transitions to `HOLEY_ELEMENTS`.
array[1] = 'b';
array[2] = 'c';
// At this point, all three positions in the array are filled, so
// the array is packed (i.e. no longer sparse). However, we cannot
// transition to a more specific kind such as `PACKED_ELEMENTS`. The
// elements kind remains `HOLEY_ELEMENTS`.
```

https://v8.dev/blog/elements-kinds#avoid-creating-holes

Reply:
Hum, I thought that if we filled in the elements in order, we wouldn't be triggering holes. Do you know if Firefox' engine behaves similarly? If yes, we should probably also revisit how we initialize `Array`s.

Reply:
Thanks for the response.

> Do you know if Firefox' engine behaves similarly?

Not really, no. I spent some time staring at SpiderMonkey sources. They do have a `NON_PACKED` flag and like V8 it _seems_ like it only transitions to `NON_PACKED`, never in the reverse (based on usage and lack of a helper method).

```cpp
    // If this flag is not set, the elements are guaranteed to contain no hole
    // values (the JS_ELEMENTS_HOLE MagicValue) in [0, initializedLength).
    NON_PACKED = 0x80,
```

https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject.h#233

```cpp
  /*
   * Number of initialized elements. This is <= the capacity, and for arrays
   * is <= the length. Memory for elements above the initialized length is
   * uninitialized, but values between the initialized length and the proper
   * length are conceptually holes.
   */
  uint32_t initializedLength;
```

https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject.h#278

```cpp
  void markNonPacked() { flags |= NON_PACKED; }
```
https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject.h#329

```cpp
inline void NativeObject::markDenseElementsNotPacked() {
  MOZ_ASSERT(is<NativeObject>());
  getElementsHeader()->markNonPacked();
}
```
https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject-inl.h#70

One place such a transition may happen is here.

```cpp
inline void NativeObject::ensureDenseInitializedLength(uint32_t index,
                                                       uint32_t extra) {
  // Ensure that the array's contents have been initialized up to index, and
  // mark the elements through 'index + extra' as initialized in preparation
  // for a write.
```

https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject-inl.h#299

From there, tracing gets a lot harder, sorry. However based on the fact that `initializedLength` is tracked your assertion seems very plausible.

> Hum, I thought that if we filled in the elements in order, we wouldn't be triggering holes.

Reply:
Thanks for the trace. The `initializedLength` does suggest that I am right for Firefox, indeed. I'll keep this in mind, and maybe one day do benchmarks when we change how `Array`s are initialized.

Reply:
One last follow-up: I installed d8 (debug v8) as suggested in the linked blog post.

```
$ ~/.jsvu/bin/v8-debug 
V8 version 11.1.192
d8> 
```

I was able to replicate their example:

```js
const array = [1, 2, 3];
array[3] = 4.56;
```

```
$ ~/.jsvu/bin/v8-debug --trace-elements-transitions test.js 
elements transition [PACKED_SMI_ELEMENTS -> PACKED_DOUBLE_ELEMENTS] in ~+15 at test.js:1 for 0x0e520010b461 <JSArray[3]> from 0x0e520025a3c1 <FixedArray[3]> to 0x0e520010b471 <FixedDoubleArray[22]>
```

But if I initialize elements one-by-one, then the initial state becomes `HOLEY_SMI_ELEMENTS`.

```js
const array = new Array(3);
array[0] = 1;
array[1] = 2;
array[2] = 3;
array[3] = 4.56;
```

```
elements transition [HOLEY_SMI_ELEMENTS -> HOLEY_DOUBLE_ELEMENTS] in ~+60 at test.js:1 for 0x02d60010b48d <JSArray[3]> from 0x02d60010b49d <FixedArray[3]> to 0x02d60010b4b1 <FixedDoubleArray[22]>
```

So unfortunately this is a sub-optimal initialization pattern on V8.